### PR TITLE
Task 113223 integrate wp with cosmosdb

### PR DESCRIPTION
**Pull request by faprovitolo**
Fix integration tests to create WorkerLease through ServiceCollectionExtensions and to avoid them to be created manually

### CODE REVIEW

File: src/Eshopworld.WorkerProcess/Configuration/Autofac/ContainerBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        /// Adds <see cref="DistributedLock"/> to the services collection
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="configuration"></param>
+        public static void AddCosmosDistributedLock(this ContainerBuilder builder, IConfigurationRoot configuration)
+        {
+            var cosmosDataStoreOptions = configuration.BindSection<CosmosDataStoreOptions>(CosmosDataStoreOptions,
+                m => { m.AddMapping(x => x.ConnectionString, "cm--cosmos-connection--worker-process"); });
+
+            builder.Register(ctx => Options.Create(cosmosDataStoreOptions))
```
Comment (tabman83) [inline src/Eshopworld.WorkerProcess/Configuration/Autofac/ContainerBuilderExtensions.cs:24]: We are registering this just to resolve it later in line 33, can we not just inject it there?

**Review by tabman83 — CHANGES_REQUESTED** [review]

File: src/Eshopworld.WorkerProcess/Configuration/ServiceConfigurationExtensions.cs [hunk 2]
```diff
@@ -16,55 +17,70 @@ namespace EShopworld.WorkerProcess.Configuration
...
-
-            var options = serviceProvider
-                .GetService<IOptionsMonitor<WorkerLeaseOptions>>();
+            services.AddOptions();
+            services.Configure<WorkerLeaseOptions>(configuration.GetSection(WorkerLeaseOptions));
 
-            if (options.CurrentValue == null)
-                throw new WorkerLeaseException($"[{typeof(IOptionsMonitor<WorkerLeaseOptions>)}] was not configured");
+            var cosmosDataStoreOptions = configuration.BindSection<CosmosDataStoreOptions>(CosmosDataStoreOptions,
+                m => { m.AddMapping(x => x.ConnectionString, "cm--cosmos-connection--worker-process"); });
```
Comment (tabman83) [inline src/Eshopworld.WorkerProcess/Configuration/ServiceConfigurationExtensions.cs:34]: can we put the secret name in a constant and reuse it

File: src/Eshopworld.WorkerProcess/Configuration/ServiceConfigurationExtensions.cs [hunk 3]
```diff
@@ -16,55 +17,70 @@ namespace EShopworld.WorkerProcess.Configuration
...
-        /// <param name="cosmosDbConnectionOptions"></param>
-        public static void AddCosmosDistributedLock(this ContainerBuilder builder, CosmosDataStoreOptions cosmosDataStoreOptions, CosmosDbConnectionOptions cosmosDbConnectionOptions)
+        /// <param name="services"></param>
+        /// <param name="configuration"></param>
+        public static void AddCosmosDistributedLock(this IServiceCollection services, IConfigurationRoot configuration)
         {
-            builder.Register(ctx => Options.Create(cosmosDataStoreOptions))
-                .As<IOptions<CosmosDataStoreOptions>>().SingleInstance();
+            var cosmosDataStoreOptions = configuration.BindSection<CosmosDataStoreOptions>(CosmosDataStoreOptions,
+                m => { m.AddMapping(x => x.ConnectionString, "cm--cosmos-connection--worker-process"); });
```
Comment (tabman83) [inline src/Eshopworld.WorkerProcess/Configuration/ServiceConfigurationExtensions.cs:55]: see comment above

File: src/Eshopworld.WorkerProcess/Configuration/Autofac/ContainerBuilderExtensions.cs (hunk 1 above)
Comment (faprovitolo) [reply src/Eshopworld.WorkerProcess/Configuration/Autofac/ContainerBuilderExtensions.cs:24]: We can certainly use it in line 33, but registration is required by DistributedLock

File: src/Eshopworld.WorkerProcess/Configuration/ServiceConfigurationExtensions.cs (hunk 2 above)
Comment (faprovitolo) [reply src/Eshopworld.WorkerProcess/Configuration/ServiceConfigurationExtensions.cs:34]: absolutely

File: src/Eshopworld.WorkerProcess/Configuration/Autofac/ContainerBuilderExtensions.cs (hunk 1 above)
Comment (tabman83) [reply src/Eshopworld.WorkerProcess/Configuration/Autofac/ContainerBuilderExtensions.cs:24]: I  am not getting you.
I mean to remove line 24 and change line 30 to look like this:
```
builder.Register(ctx =>
    new CosmosDistributedLockStore(
        documentClient,
        Options.Create(cosmosDataStoreOptions),
        ctx.Resolve<IBigBrother>())).As<ICosmosDistributedLockStore>().SingleInstance();

```

Comment (faprovitolo) [reply src/Eshopworld.WorkerProcess/Configuration/Autofac/ContainerBuilderExtensions.cs:24]: ok but how will IOptions<CosmosDataStoreOptions> be resolved in DistributedLock constructor? If we remove registration?

Comment (tabman83) [reply src/Eshopworld.WorkerProcess/Configuration/Autofac/ContainerBuilderExtensions.cs:24]: Ah yes sorry I missed that on the last line!

File: src/Eshopworld.WorkerProcess/Configuration/ServiceConfigurationExtensions.cs (hunk 2 above)
Comment (faprovitolo) [reply src/Eshopworld.WorkerProcess/Configuration/ServiceConfigurationExtensions.cs]: fixed

**Review by tabman83 — APPROVED** [review]

**Pull request merged**